New script for making book repositories public

TylerZeroMaster opened the pull request:
The python script is he main entry point and will automatically install any additional dependencies. Note that it will install `ts-node` globally. I used python instead of bash because I prefer python's error handling. I stole most of the `GitRepo` class from a personal project and added support for remotes.

All of the python functions and classes that use the GitHub CLI (`gh`) are not used anymore. I kept them around because they might be useful again. For example, today I used `list_prs` to help me find all the books that had PRs. There was such a small number of PRs that I decided to manually close them. 

I used typescript for generating the README and copying the settings because we would like to have this functionality in POET eventually. Hopefully it will be fairly easy to migrate this script.

TylerZeroMaster committed:
New script for making book repositories public

philschatz commented on git-repo-prep/static/book.md:
```diff
@@ -0,0 +1,11 @@
+# {book_title}
```
Could the README have a Gitpod badge?

TylerZeroMaster replied:
If you mean like this:
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/openstax/osbooks-business-law)
It should not be too difficult to add.

philschatz replied:
@TylerZeroMaster I think you can use the referrer URL inside the README instead of templating the repo name

philschatz commented on git-repo-prep/static/repo-settings/.github/settings.yml:
```diff
@@ -0,0 +1,45 @@
+  - name: ce-all
+    permission: triage
+  - name: content-managers
+    permission: admin
```
I thought that CM's (&ce-admin) no longer need admin permission because updating settings will be done by updating this file

philschatz commented on git-repo-prep/static/repo-settings/.github/settings.yml:
```diff
@@ -0,0 +1,45 @@
+  has_projects: false
+  has_downloads: false
+
+  default_branch: main
```
Does this settings.yml file contain all of the options that the probot settings plugin understands? (just checking. Hopefully it does)

I'm torn... It'd be nice to see which of these options are defaults and which were changed by humans. I guess git blame can help point that out. hmm.

TylerZeroMaster replied:
I did not include the `labels` or `milestones` configuration sections.

The `repository` section is missing these:
```yml
  allow_squash_merge: true
  allow_merge_commit: true
  allow_rebase_merge: true

  enable_automated_security_fixes: false
  enable_vulnerability_alerts: false

  topics: github, probot
  homepage: ...
  name: ...
```

Other than that, everything that I know of that could be included is included.

philschatz reviewed (commented):
Just added some notes. We can :parking: it tomorrow if that's easier

TylerZeroMaster commented on git-repo-prep/main.py:
```diff
@@ -0,0 +1,302 @@
+            self('add .')
+            self(f'commit -m "{msg}"')
+
+    def push_changes(self):
```
💡

TylerZeroMaster committed:
Add POET and use cli to find orphans to remove

This needs to be revisited because there cases where it is not clear
if it is okay to remove certain files.

Right now, POET considers everything in the Interactive directory to be
orphaned.

POET also considers text files like
osbooks-principles-of-management-bundle/media/4-20-longdesc.txt to be
orphaned.

For now, we will just check the root directory of the book for orphans.

Also fixed a bug with tag and branch output parsing (it would return
[''], now it will not)

TylerZeroMaster committed:
Improve the handling of license in index.ts and fix PATH issues

Reduce likelihood license is replaced needlessly

Verify that all licenses match in a bundle using ts instead of python

philschatz commented on git-repo-prep/main.py:
```diff
@@ -0,0 +1,302 @@
+    if not license_file.exists():
+        logging.warning(f'Unknown license: {license_file}')
+        return
+    shutil.copy(license_file, book_path/'LICENSE')
```
Here's the list that GitHub detects: https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository

TylerZeroMaster committed:
Improve README generation and trim some extraneous settings

- Encode book slug in url, update templates, and
make template placeholder syntax more like other template engines.

- Replace book_slug with book_link in book replacments map. This is
more clear and more similar to bundles.

- Set git user to staxly in main.py#init and remove some extraneous
settings from settings.yml

- remove allow_forking because it does not need to be included
in settings.yml.

philschatz commented on git-repo-prep/static/repo-settings/.github/settings.yml:
```diff
@@ -0,0 +1,27 @@
+
+  default_branch: main
+
+  allow_auto_merge: true
```
Do these need to be included? I thought they were all true by default when a new repo is created.

Similarly, I think the default_branch is main when a new repo is created, no?

TylerZeroMaster replied:
`allow_auto_merge` has a default value of `false` on private repositories. `delete_branch_on_merge` has a default value of `false` regardless of repository visibility.

`allow_forking`, however, is [completely irrelevant](https://docs.github.com/en/enterprise-server@3.5/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-forking-policy-for-your-repository) to public repositories. I removed it from the settings.

philschatz commented on git-repo-prep/index.ts:
```diff
@@ -0,0 +1,214 @@
+      slugsMeta.map(s =>
+        `- _${s.colMeta.title}_ [online](${BOOK_WEB_ROOT}${encodeURIComponent(s.slugName)})`
+      ).join('\n'),
+    'license_text': slugsMeta[0].colMeta.license.text,
```
This seems to assume that all books in a repo have the same license. I'm not sure if that is true. Could this error if the licenses are different?

Maybe the python code handles this? I see a line with the following:

```python
raise Exception('Licenses differ between collections')
```

TylerZeroMaster replied:
I figured it was better to have the check closer to the assumption, so I moved it into the typescript.

philschatz reviewed (approved):
Looks good! Want to pair up while running it?

TylerZeroMaster committed:
Final update: fix a bug, add a dry_run mode, and update README

- The default mode is dry run, users will need to read the directions
to know how to push changes.

- Branch specs are now split at most one time on / because apparently
branch names can include / (i.e. origin/my/branch/name is valid)
